oxidation/libparsec/crates/core_fs/src/storage/chunk_storage.rs

```diff
@@ -0,0 +1,409 @@
+use parsec_api_types::{ChunkID, DateTime, DEFAULT_BLOCK_SIZE};
+
+use super::local_database::{SqliteConn, SQLITE_MAX_VARIABLE_NUMBER};
+use crate::error::{FSError, FSResult};
+use crate::extensions::CoalesceTotalSize;
+use crate::schema::chunks;
+
+#[derive(Insertable, AsChangeset)]
+#[diesel(table_name = chunks)]
+struct NewChunk<'a> {
+    pub chunk_id: &'a [u8],
+    pub size: i32,
+    pub offline: bool,
+    pub accessed_on: Option<f32>,
+    pub data: &'a [u8],
+}
+
+pub(crate) trait ChunkStorageTrait {
+    fn conn(&self) -> &Mutex<SqliteConn>;
+    fn local_symkey(&self) -> &SecretKey;
+
+    fn create_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query(
+            "CREATE TABLE IF NOT EXISTS chunks (
+                chunk_id BLOB PRIMARY KEY NOT NULL, -- UUID
+                size INTEGER NOT NULL,
+                offline BOOLEAN NOT NULL,
+                accessed_on REAL, -- Timestamp
+                data BLOB NOT NULL
+            );",
+        )
+        .execute(conn)
+        .map_err(|e| FSError::CreateTable(format!("chunks {e}")))?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    fn drop_db(&self) -> FSResult<()> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        sql_query("DROP TABLE IF EXISTS chunks;")
+            .execute(conn)
+            .map_err(|e| FSError::DropTable(format!("chunks {e}")))?;
+        Ok(())
+    }
+
+    // Size and chunks
+
+    fn get_nb_blocks(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .first(conn)
+            .map_err(|e| FSError::QueryTable(format!("chunks: get_nb_blocks {e}")))
+    }
+
+    fn get_total_size(&self) -> FSResult<i64> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(CoalesceTotalSize::default())
+            .first(conn)
+            .map_err(|e| FSError::QueryTable(format!("chunks: get_total_size {e}")))
+    }
+
+    fn is_chunk(&self, chunk_id: ChunkID) -> FSResult<bool> {
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        chunks::table
+            .select(count_star())
+            .filter(chunks::chunk_id.eq((*chunk_id).as_ref()))
+            .first::<i64>(conn)
+            .map_err(|e| FSError::QueryTable(format!("chunks: is_chunk {e}")))
+            .map(|res| res > 0)
+    }
+
+    fn get_local_chunk_ids(&self, chunk_ids: &[ChunkID]) -> FSResult<Vec<ChunkID>> {
+        let bytes_id_list = chunk_ids
+            .iter()
+            .map(|chunk_id| (**chunk_id).as_ref())
+            .collect::<Vec<_>>();
+
+        let mut res = Vec::with_capacity(chunk_ids.len());
+
+        let conn = &mut *self.conn().lock().expect("Mutex is poisoned");
+        for bytes_id_list_chunk in bytes_id_list.chunks(SQLITE_MAX_VARIABLE_NUMBER) {
```

Comment:
Can you add a comment to explain what's going on here: the fact we are doing a query in a for loop seems like poor code where it is something we have really think about !

You can also mention:
```
The number of loop iteration is expected to be rather low: typically considering 4ko per chunk (i.e. the size of the buffer the Linux Kernel send us through Fuse), each query could perform ~4mo of data.

If performance ever becomes an issue, we could further optimize this by having the caller filter the chunks: a block containing multiple chunk means those correspond to local modification and hence can be filtered out. With this we would only provide the chunks corresponding to a synced block which are 512ko big, so each query performs up to 512Mo !
```